### PR TITLE
more detail in log markers in Thrall

### PR DESCRIPTION
**Pull request by AWare**
## What does this change?

Every place where previously we'd created a new log marker in thrall, I've created a new `MarkerContext` and passed that down. 

We create new `MarkerContext`s from an existing (explicit) `MarkerContext` and a `Map[String,Any]`- use the map to create a brand new Marker, add any Marker present in the existing context and then create and return a new `MarkerContext`. 

We've agreed that this should be as explicit as possible, so I'm going through and making it so.

I think it's currently functional, but I'm going to go and make it explicit when it creates the initial context from the message. 

Where we have a function which doesn't do anything to a `MarkerContext` I've just left it implicit. 
 
## How can success be measured?

More details in logs!

## Tested?
- [ ] locally
- [ ] on TEST

### CODE REVIEW

File: thrall/app/lib/RetryHandler.scala [hunk 1]
```diff
@@ -43,12 +44,13 @@ object RetryHandler {
...
     }
 
-    def handleWithRetry[T](f: () => Future[T], retries: Int, delay: FiniteDuration): () => Future[T] = () => {
+    def handleWithRetry[T](f: () => Future[T], retries: Int, delay: FiniteDuration)(implicit markerContext: MarkerContext): () => Future[T] = () => {
       implicit val scheduler = actorSystem.scheduler
       var count = 0
 
       def attempt = () => {
         count = count + 1
+        MarkerAugmentation.augmentMarkerContext(Map("attempt" -> count))
```
Comment (jonathonherbert) [inline thrall/app/lib/RetryHandler.scala]: Could we be more explicit here?

```suggestion
        MarkerAugmentation.augmentMarkerContext(Map("retryCount" -> count))
```

File: thrall/app/lib/RetryHandler.scala [hunk 2]
```diff
@@ -3,24 +3,27 @@ package lib
...
 import akka.pattern.{after, retry}
+import com.gu.mediaservice.lib.logging.MarkerAugmentation
 import play.api.{Logger, MarkerContext}
 
 import scala.concurrent.{ExecutionContext, Future, TimeoutException}
 import scala.concurrent.duration.FiniteDuration
 import scala.util.{Failure, Success}
 
 object RetryHandler {
+  def withExplicitMarkerContext[T](f: () => Future[T], mc: MarkerContext): Future[T] = {
```
Comment (AWare) [inline thrall/app/lib/RetryHandler.scala]: This doesn't give me good feelings.

File: common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerAugmentation.scala [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  /** *
+    *
+    * @param markerContext
+    * @param augmentation : more details!
+    * @return a new markercontext!
+    *         it wouldn't be java if it could be cloned
+    */
+  def augmentMarkerContext(markerContext: MarkerContext, augmentation: Map[String, Any]): MarkerContext = {
+    val newMarker = appendEntries(augmentation.asJava)
+    markerContext.marker.foreach(newMarker.add)
```
Comment (akash1810) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerAugmentation.scala]: What's the value of `markerContext.marker` here?

File: common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerAugmentation.scala [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    MarkerContext(newMarker)
+  }
+  def augmentMarkerContext(markerContext: MarkerContext, augmentations: (String, Any)*): MarkerContext = {
```
Comment (akash1810) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerAugmentation.scala]: What scenario would we use this function over the one above?

File: common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerAugmentation.scala (hunk 3 above)
Comment (AWare) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerAugmentation.scala]: `markerContext.marker` is an `Option[Marker]`

File: common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerAugmentation.scala (hunk 4 above)
Comment (AWare) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerAugmentation.scala]: it's just sugar

File: common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerUtils.scala [hunk 5]
```diff
@@ -2,15 +2,28 @@ package com.gu.mediaservice.lib.logging
...
+
+
+case class MarkerMap(map: Map[String, Any]) extends LogMarker {
+  override def toLogMarker: LogstashMarker = appendEntries(map.asJava)
+}
 
-case class MarkerMap(values: Map[String, Any]) extends LogMarker {
-  override def toLogMarker: LogstashMarker = appendEntries(values.asJava)
+object MarkerMap {
+  def apply(entries: (String, Any)*):MarkerMap = MarkerMap(entries.toMap)
```
Comment (akash1810) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerUtils.scala]: NICE!

File: common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerUtils.scala [hunk 6]
```diff
@@ -2,15 +2,28 @@ package com.gu.mediaservice.lib.logging
...
 import net.logstash.logback.marker.Markers.appendEntries
+import play.api.MarkerContext
+import scala.language.implicitConversions
+
 import scala.collection.JavaConverters._
 
-trait LogMarker { def toLogMarker: LogstashMarker }
+trait LogMarker {
+  def toLogMarker: LogstashMarker
+  def map: Map[String, Any]
```
Comment (akash1810) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerUtils.scala]: can we name this to be more intention revealing?

File: kahuna/package.json [hunk 7]
```diff
@@ -50,7 +50,7 @@
     "test": "npm run lint && jest --coverage",
     "lint": "eslint public/js && htmllint public/js/**/*.html",
     "lint-fix": "eslint public/js --fix && htmllint public/js/**/*.html",
-    "dist": "webpack --progress --config webpack.config.prod.js",
+    "dist": "webpack --config webpack.config.prod.js",
```
Comment (akash1810) [inline kahuna/package.json]: Let's make this it's own PR?

File: dev-start.sh [hunk 8]
```diff
@@ -116,8 +116,8 @@ setupLocalKinesis() {
   # ignore stream already exists error
   set +e
   stream_name='media-service-DEV-ThrallMessageQueue-1N0T2UXYNUIC9'
-  aws --profile media-service --region=eu-west-1 --endpoint-url=http://localhost:4568 kinesis create-stream --shard-count 1 --stream-name "${stream_name}"
-  aws --profile media-service --region=eu-west-1 --endpoint-url=http://localhost:4568 kinesis list-streams
+  aws --profile media-service --region=eu-west-1 --endpoint-url=http://localhost:4568 kinesis create-stream --shard-count 1 --stream-name "${stream_name}" | cat
```
Comment (akash1810) [inline dev-start.sh]: Let's make this it's own PR? https://github.com/guardian/grid/pull/2867

File: project/plugins.sbt [hunk 9]
```diff
@@ -1,4 +1,4 @@
-resolvers += "Typesafe repository" at "http://repo.typesafe.com/typesafe/releases/"
+resolvers += "Typesafe repository" at "https://repo.typesafe.com/typesafe/releases/"
```
Comment (akash1810) [inline project/plugins.sbt]: YES!!

File: scripts/ci/grid.sh [hunk 10]
```diff
@@ -13,6 +13,8 @@ setupNvm() {
 }
 
 buildJs() {
+  echo "##teamcity[compilationStarted compiler='webpack']"
```
Comment (akash1810) [inline scripts/ci/grid.sh]: NICE!

File: thrall/app/lib/RetryHandler.scala [hunk 11]
```diff
@@ -3,57 +3,73 @@ package lib
...
                                     executionContext: ExecutionContext,
-                                    mc: MarkerContext
-                                  ): () => Future[T] = {
-    def logFailures[T](f: () => Future[T])(
-      implicit executionContext: ExecutionContext, mc: MarkerContext
-    ): () => Future[T] = {
-      () => {
-        f().transform {
+                                  ): Future[T] = {
+    println("ALEX THREE")
```
Comment (akash1810) [inline thrall/app/lib/RetryHandler.scala]: let's remove the `println`s

File: common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerUtils.scala (hunk 6 above)
Comment (AWare) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerUtils.scala]: it's now called `markerContents`

**Comment by AWare** [comment]
Dotting i's and crossing t's

- [ ] image
- [ ] reindex-image
- [ ] delete-image
- [ ] update-image
- [ ] delete-image-exports
- [ ] update-image-exports
- [ ] update-image-user-metadata
- [ ] update-image-usages
- [ ] replace-image-leases
- [ ] add-image-lease
- [ ] remove-image-lease
- [ ] set-image-collections
- [ ] delete-usages
- [ ] upsert-rcs-rights
- [ ] update-image-photoshoot
- [ ] batch-index

File: common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerUtils.scala [hunk 12]
```diff
@@ -2,15 +2,28 @@ package com.gu.mediaservice.lib.logging
...
 import net.logstash.logback.marker.LogstashMarker
 import net.logstash.logback.marker.Markers.appendEntries
+import play.api.MarkerContext
+import scala.language.implicitConversions
+
 import scala.collection.JavaConverters._
 
-trait LogMarker { def toLogMarker: LogstashMarker }
+trait LogMarker {
+  def toLogMarker: LogstashMarker
```
Comment (akash1810) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerUtils.scala]: The implementation of this is the same (other than in `MarkerMap`); is it worth providing a default concrete implementation in the trait which can be overridden if you need to something different?

File: common-lib/src/main/scala/com/gu/mediaservice/lib/elasticsearch/ElasticSearchException.scala [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+import net.logstash.logback.marker.LogstashMarker
+
+trait ElasticSearchError {
+  self: Throwable =>
+  def error: ElasticError
+
+  def markerContents: Map[String, Any]
+}
+
+object ElasticSearchException {
```
Comment (akash1810) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/elasticsearch/ElasticSearchException.scala]: does this need to extend `LogMarker`?

Comment (AWare) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/elasticsearch/ElasticSearchException.scala]: No, because it has an unapply that gives us a `LogMarker`

File: common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerUtils.scala (hunk 12 above)
Comment (AWare) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/logging/MarkerUtils.scala]: Yeah that's a good idea.

**Comment by AWare** [comment]
Put this on TEST, and it's working for me. 

Not sure how to trigger the remaining messages, would like to just double check them once live.
